### PR TITLE
MOL-3617: Feat/hypervalency

### PR DESCRIPTION
**Pull request by grahamtholt**
[MOL-3617](https://linear.app/vant/issue/MOL-3617/implement-metric-for-hypervalency)

Add reference-free metric to count the number of hypervalent atoms (~~ atoms that throw RDKit `ExplicitValenceError`). To do this I:

- Added a metric for counting the number of hypervalent atoms. This uses RDKit under the hood.

~NB: There is some ambiguity on whether to consider an atom as "hypervalent" and therefore invalid or simply charged. One of my test cases ("CO(C)C") should probably be considered hypervalent, but is currently not counted as such. Could use some input on this @padix-key @OleinikovasV~

### CODE REVIEW

File: tests/test_metric_reffree.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    one_viol = {Chem.MolFromSmiles(smile, params=params): 1 for smile in one_viol}
+    two_viols = {Chem.MolFromSmiles(smile, params=params): 2 for smile in two_viols}
+    return good_mols | one_viol | two_viols
+
+
+@pytest.mark.parametrize(
+    ["mol", "expected_viols"],
+    [(mol, num_exp_viols) for mol, num_exp_viols in _test_mols().items()],
+)
+def test_count_valence_violations(mol, expected_viols):
```
Comment (padix-key) [inline tests/test_metric_reffree.py:42]: Could you add docstrings to these test functions, that briefly explain what should be tested and how it is done?
```suggestion
def test_count_valence_violations(mol, expected_viols):
"""
Check if the correct number of valence violations are found by
:class:`LigandValenceViolations` by checking the metric against molecules
with a known number of violations.
"""
```

File: tests/test_metric_reffree.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        "C1=CC=CC=C1[H]",
+        "N(=O)=O",
+        "C=P(=O)O",
+        "O=Cl(=O)O",
+        "OC1C(COP(=O)(OP(=O)(OP(=O)(O)O)O)O)OC(C1O)n1cnc2c1ncnc2N",
+    ]
+
+    one_viol = [
+        "C[C@H](FC)C",
+        "CO(C)C",  # This probably shouldn't read as 0 violations, but it does currently.
```
Comment (padix-key) [inline tests/test_metric_reffree.py:23]: I think for now it is OK to have permissive edge cases, which are not counted as violations when they probably should. Independent of the implementation in `RDKit` it would also be hard to tell for me when some seemingly wild atomic charges are actually valid, i.e. I think there are not really strict simple-to-implement rules for these cases, are there?

**Review by padix-key — COMMENTED** [review]
Hi, thanks for the implementation. I left a few suggestions and questions below.

File: src/peppr/system_subset.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    def select(self, pose: struc.AtomArray) -> npt.NDArray[np.bool_]:
+        ligand_mask = pose.hetero
+        chain_starts = struc.get_chain_starts(pose)
+        if len(chain_starts) == 0:
+            # No chains in the structure
+            return np.array([], dtype=bool)
+        chain_masks = struc.get_chain_masks(pose, chain_starts)
+        # Only keep chain masks that correspond to ligand chains
+        ligand_masks = chain_masks[(chain_masks & ligand_mask).any(axis=-1)]
+        return ligand_masks
```
[the remark this answers is not in the document]
Comment (grahamtholt) [reply src/peppr/system_subset.py:48]: `select` is specced to return a single numpy array containing multiple masks, where the first dimension of the array indexes the masks.

```
        Returns
        -------
        NDArray[bool], shape=(n_subsets, n)
            The masks that can be used to select subsets of the system.
            The first dimension of the array is the subset index.
```

File: src/peppr/metric_reffree.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+from typing import OrderedDict
+import biotite.interface.rdkit as rdkit_interface
+import biotite.structure as struc
+import numpy as np
+from rdkit import Chem
+from peppr.sanitize import sanitize
+from peppr.system_subset import IsolatedLigandSelector
+
+
+class RefFreeMetric(ABC):
```
[the remark this answers is not in the document]
Comment (grahamtholt) [reply src/peppr/metric_reffree.py:14]: In that case would we want to modify the signature of `Metric.evaluate` to make the reference optional? I.e., 

```python
    def evaluate(self, pose: struc.AtomArray, reference: struc.AtomArray | None) -> float:
```

File: tests/test_metric.py [hunk 5]
```diff
@@ -424,3 +429,178 @@ def _assemble_matched_predictions(system_id):
...
+
+
+@pytest.mark.parametrize(
+    ["mol", "expected_viols"],
+    [
+        (mol, num_exp_viols)
+        for mol, num_exp_viols in _test_mols_valence_violations().items()
+    ],
+)
+def test_count_valence_violations(mol, expected_viols):
```
Comment (padix-key) [inline tests/test_metric.py:467]: I would be a but cautious testing private functions, because they require the reader to know more than the API and because private functions change more frequently than the relatively stable public API. I usually only resort to this, if some functionality cannot be reasonably tested using the public API.

In this case however, the test below achieves basically the same, as `LigandValenceViolations.evaluate()` is only a thin wrapper around `_count_valence_violations()`. Therefore I think we can remove this test.
